File: data_structures/trees/Binary Tree Traversal/binary_search_tree.cpp



/*
Tree traversal

In computer science, tree traversal (also known as tree search) is a form of graph traversal and refers
to the process of visiting (checking and/or updating) each node in a tree data structure, 
exactly once. Such traversals are classified by the order in which the nodes are visited. 
The following algorithms are described for a binary tree, but they may be generalized to other
trees as well.

There are two categories of traversals:
1. Depth First Traversals:
(a) Inorder (Left, Root, Right)
(b) Preorder (Root, Left, Right)
(c) Postorder (Left, Right, Root)

2. Breadth First Traversal (Or Level Order Traversal)
    In Breadth First Traversal, the root node is visited first, then the nodes on the next level, then the nodes on the next level, and so on.
*/

#include<iostream>
using namespace std;

struct Node{
    int data;
    struct Node *left, *right;
};

struct Node* newNode(int data){
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

Node* head = NULL;

void insert(int data){
    if(head == NULL){
        head = newNode(data);
        return;
    }
    if(data <= head->data){
        head = head->left;
    }else{
        head = head->right;
    }
    insert(data);
}

void search(int x){
    if(head == NULL){

    }
}

int main( int argc, char *argv[] ){
/*
    Given an array of integers the program
    should print array in inorder, preorder and postorder traversal.
*/

        // 1, 4, 2, 3, 5, 6, 7
             /*
              head
                 \
                 1
              /      \
            NULL      4
                    /
                   2
                    \
                     3
                      \
                       5
                        \
                         6
                          \
                           7
                            \
                            NULL
                */
    for(int i=1; i < argc; i++){
        int data = atoi(argv[i]);
        insert(data);
    }

    while(head != NULL){
        cout << head->data << " ";
        head = head->right;
    }
    return 0;
}

File: data_structures/trees/binary_search_tree.py
import sys
sys.path.append('../trees')  # Add the correct path to the sys.path list
from node import Node  # Import the 'Node' class from the correct package

class Tree:
    def __init__(self):
        self.root = None

    def binary_search(self, root, key):
        if root is None:
            return False
        if root.data == key:
            return True
        if root.data < key:
            return self.binary_search(root.right, key)
        return self.binary_search(root.left, key)

# Tree structure
#           9
#         /   \
#        3     12
#       / \   / \
#      1   4 10  17
#     / \      \   \
#    0   2     11   19
#
if __name__ == '__main__':
    tree = Tree()
    tree.root = Node(9)
    tree.root.left = Node(3)
    tree.root.right = Node(12)
    tree.root.left.left = Node(1)
    tree.root.left.right = Node(4)
    tree.root.right.left = Node(10)
    tree.root.right.right = Node(17)
    tree.root.left.left.left = Node(0)
    tree.root.left.left.right = Node(2)
    tree.root.right.left.right = Node(11)
    tree.root.right.right.right = Node(19)
    print("search for 17 ", tree.binary_search(tree.root, 17)) # True
    print("search for 10", tree.binary_search(tree.root, 10)) # True
    print("search for 1 ", tree.binary_search(tree.root, 1)) # True
    print("search for 9 ", tree.binary_search(tree.root, 9)) # True
    print("search for 8 ", tree.binary_search(tree.root, 8)) # False
    print("search for 6 ", tree.binary_search(tree.root, 6)) # False
    print("search for 12 ", tree.binary_search(tree.root, 12)) # True

